src/pepper/parser.py

```diff
@@ -81,18 +99,304 @@ def p_include_expression(p: yacc.YaccProduction) -> yacc.YaccProduction:
+    '''
+
+    p[0] = p[1] << p[3]
+
+def p_valid_bit_rshift(p: yacc.YaccProduction) -> yacc.YaccProduction:
+    '''
+    valid_expr : valid_expr R_SHIFT valid_expr
+    '''
+
+    p[0] = p[1] >> p[3]
+
+def p_valid_logic_or(p: yacc.YaccProduction) -> yacc.YaccProduction:
+    '''
+    valid_expr : valid_expr BOOL_OR valid_expr
+    '''
+
+    p[0] = bool(p[1] or p[3])
+
+def p_valid_logic_and(p: yacc.YaccProduction) -> yacc.YaccProduction:
+    '''
+    valid_expr : valid_expr BOOL_AND valid_expr
+    '''
+
+    p[0] = bool(p[1] and p[3])
+
+def p_valid_logic_not(p: yacc.YaccProduction) -> yacc.YaccProduction:
+    '''
+    valid_expr : spaces '!' spaces valid_expr
+    '''
+
+    p[0] = (not p[4] ) == True
+
+def p_valid_logic_lt(p: yacc.YaccProduction) -> yacc.YaccProduction:
+    '''
+    valid_expr : valid_expr '<' valid_expr
+    '''
+    p[0] = p[1] < p[3]
+
+def p_valid_logic_le(p: yacc.YaccProduction) -> yacc.YaccProduction:
+    '''
+    valid_expr : valid_expr COMP_LTE valid_expr
+    '''
+    p[0] = p[1] <= p[3]
+
+def p_valid_logic_gt(p: yacc.YaccProduction) -> yacc.YaccProduction:
+    '''
+    valid_expr : valid_expr '>' valid_expr
+    '''
+    p[0] = p[1] > p[3]
+
+def p_valid_logic_ge(p: yacc.YaccProduction) -> yacc.YaccProduction:
+    '''
+    valid_expr : valid_expr COMP_GTE valid_expr
+    '''
+    p[0] = p[1] >= p[3]
+
+def p_valid_equal(p: yacc.YaccProduction) -> yacc.YaccProduction:
+    '''
+    valid_expr : valid_expr COMP_EQU valid_expr
+    '''
+    p[0] = p[1]  == p[3]
+
+def p_valid_nequal(p: yacc.YaccProduction) -> yacc.YaccProduction:
+    '''
+    valid_expr : valid_expr COMP_NEQU valid_expr
+    '''
+    p[0] = p[1] != p[3]
+
+def p_valid_ternary(p: yacc.YaccProduction) -> yacc.YaccProduction:
+    '''
+    valid_expr : valid_expr spaces '?' spaces valid_expr ':' valid_expr
+    '''
+    p[0] = p[5] if p[1] else p[7]
+
+
+def p_if_expression(p: yacc.YaccProduction) -> yacc.YaccProduction:
+    """
+    if_expression : PREPROCESSING_KEYWORD_IF WHITESPACE valid_expr
+    """
+    global if_count
+    if_count += 1
+
+    symtable.IF_STACK.append((str(if_count), True))
+    p[0] = ast.StringLiteralNode([f"// if expression result: { int(p[3]) }"])
+
+
+
+def p_no_space(p: yacc.YaccProduction) -> yacc.YaccProduction:
+    '''
+    spaces :
+    '''
+
+    p[0] = None
+
+def p_spaces(p: yacc.YaccProduction) -> yacc.YaccProduction:
+    '''
+    spaces : WHITESPACE spaces
+    '''
+
+    p[0] = ast.WhiteSpaceNode([p[1]])
+
+# TODO: make nodes instead that have the appropriate children and evaluated expression?
+####### DONE
 
 def p_ifdef_expression(p: yacc.YaccProduction) -> yacc.YaccProduction:
     """
     ifdef_expression : PREPROCESSING_KEYWORD_IFDEF WHITESPACE IDENTIFIER
     """
-    if p[3] in symtable.TABLE.keys():
-        symtable.IFDEF_STACK.append((p[3], True))
-    else:
-        symtable.IFDEF_STACK.append((p[3], False))
+    symtable.IF_STACK.append((p[3], p[3] in symtable.TABLE.keys()))
```

Comment:
this can actually just be `symtable.TABLE`--it searches dict keys by default, AFAIK

Reply:
I knew that, I originally kept the expression like that because that was your expression in the original if statement. I thought maybe you wanted to do that for readability